### PR TITLE
Simplify higher-ranked LUB/GLB

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is a better version of https://github.com/rust-lang/rust/pull/44211. It still makes higher-ranked LUB/GLB into a hard equality test, however, it does try to identify that something changed and issue a notice to the user. I wroteup https://github.com/rust-lang/rust/issues/45852 as a tracking issue for this change.

Currently, this moves straight to a hard-error, on the basis that the crater run in #44211 saw no impact. It might be good to retest -- or perhaps to try for a warning period. Trying to do the latter in a precise way would be somewhat painful, but an imprecise way might suffice -- that is, we could issue warning *whenever* a LUB/GLB operation succeeds that will later fail, even if it doesn't ultimately impact the type check. I could experiment with this.

~~I am *mildly* wary about landing this independently of other code that moves to a universe-based system. In particular, I was nervous that this change would make coherence accepts new pairs of impls that will later be errors. I have the code for the universe-based approach available, I hope to open an PR and run some tests on its impact very shortly.~~ @arielb1 points out that I was being silly.

r? @arielb1

### CODE REVIEW

**Comment by arielb1** [comment]
> In particular, I was nervous that this change would make coherence accepts new pairs of impls that will later be errors

Coherence is based only on type equality, while this PR only affects subtyping - I don't believe the LUB code can be run from coherence.

**Comment by arielb1** [comment]
if that wasn't clear, r=me unless there's some other complication about coherence.

**Comment by kennytm** [comment]
`x86_64-windows-msvc` UI test failed:

```diff
 error[E0308]: match arms have incompatible types
   --> $DIR/old-lub-glb-object.rs:20:13
    |
 20 |       let z = match 22 {
    |  _____________^
 21 | |         0 => x,
 22 | |         _ => y,
 23 | |     };
-   | |_____^ expected bound lifetime parameter 'a, found concrete lifetime
+   | |_____^ expected bound lifetime parameter 'b, found concrete lifetime
    |
    = note: expected type `&for<'a, 'b> Foo<&'a u8, &'b u8>`
               found type `&for<'a> Foo<&'a u8, &'a u8>`
    = note: this was previously accepted by the compiler but has been phased out
    = note: for more information, see https://github.com/rust-lang/rust/issues/45852
 note: match arm with an incompatible type
   --> $DIR/old-lub-glb-object.rs:22:14
    |
 22 |         _ => y,
    |              ^
 
 error: aborting due to previous error
```

**Comment by nikomatsakis** [comment]
Argh. I wonder if this is the same problem we saw in https://github.com/rust-lang/rust/pull/44167. I'm pulling @cengizIO's commits here temporarily just to see what AppVeyor says.

**Comment by arielb1** [comment]
I think we should wait for his PR to land and then rebase this one on top of it

**Comment by kennytm** [comment]
`armhf-gnu`, [exactly the same error as before](https://github.com/rust-lang/rust/pull/45853#issuecomment-343714497).

<details>

```diff
 error[E0308]: match arms have incompatible types
   --> $DIR/old-lub-glb-object.rs:20:13
    |
 20 |       let z = match 22 {
    |  _____________^
 21 | |         0 => x,
 22 | |         _ => y,
 23 | |     };
-   | |_____^ expected bound lifetime parameter 'a, found concrete lifetime
+   | |_____^ expected bound lifetime parameter 'b, found concrete lifetime
    |
    = note: expected type `&for<'a, 'b> Foo<&'a u8, &'b u8>`
               found type `&for<'a> Foo<&'a u8, &'a u8>`
    = note: this was previously accepted by the compiler but has been phased out
    = note: for more information, see https://github.com/rust-lang/rust/issues/45852
 note: match arm with an incompatible type
   --> $DIR/old-lub-glb-object.rs:22:14
    |
 22 |         _ => y,
    |              ^
 
 error: aborting due to previous error
```

</details>

**Comment by nikomatsakis** [comment]
@kennytm thanks, I'll take a look more closely then. Thought we had squelched that.

**Comment by kennytm** [comment]
@nikomatsakis Looks like Travis passed this time.

**Comment by nikomatsakis** [comment]
@kennytm yep, thanks

**Comment by kennytm** [comment]
`x86_64-gnu-cargotest`, ICE when compiling `servo`, legit.

```
[00:56:06] error: internal compiler error: unexpected panic
[00:56:06] 
[00:56:06] note: the compiler unexpectedly panicked. this is a bug.
[00:56:06] 
[00:56:06] note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
[00:56:06] 
[00:56:06] note: rustc 1.23.0-dev running on x86_64-unknown-linux-gnu
[00:56:06] 
[00:56:06] thread 'rustc' panicked at 'index out of bounds: the len is 141 but the index is 141', /checkout/src/liballoc/vec.rs:1551:10
[00:56:06] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:56:06] 
[00:56:06] error: Could not compile `style`.
```

File: src/librustc/infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -764,16 +764,24 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
             }
         }
 
-        self.note_error_origin(diag, &cause);
         self.check_and_note_conflicting_crates(diag, terr, span);
         self.tcx.note_and_explain_type_err(diag, terr, span);
+        self.check_and_note_conflicting_crates(diag, terr, span);
```
Comment (arielb1) [inline src/librustc/infer/error_reporting/mod.rs]: duplicate call to `check_and_note_conflicting_crates`?

**Comment by nikomatsakis** [comment]
@arielb1 argh

**Comment by arielb1** [comment]
I think it's a good idea to have #45825 land first

**Comment by nikomatsakis** [comment]
Agreed. Waiting for https://github.com/rust-lang/rust/pull/45825 to land.

**Comment by kennytm** [comment]
@bors retry 

`check i686-pc-windows-msvc` 3-hour timeout

**Pull request merged**